src/sage/schemes/projective/cohomology.py

```diff
@@ -0,0 +1,500 @@
+
+On the other hand, it is not clear how to represent `H^0(\tilde M)` in terms of
+twisted structure sheaves of a projective space. Hence it is merely created
+as a vector space over `k` with the correct dimension::
+
+    sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+    sage: X = P2.subscheme([x^4 + y^4 + z^4])
+    sage: sh = X.structure_sheaf()
+    sage: sh.cohomology(0)
+    1
+    sage: sh._cohomology_group(0)
+    Vector space of dimension 1 over Rational Field
+"""
+
+from sage.misc.flatten import flatten
+from sage.combinat.integer_lists.invlex import IntegerListsLex
+from sage.modules.free_module import VectorSpace
+from sage.modules.free_module_element import vector
+
+
+class CohomologyGroupBottom:
+    r"""
+    Bottom cohomology group of the twisted structure sheaf of a projective space.
+
+    INPUT:
+
+    - ``S`` -- a direct sum of copies of the coordinate ring of a projective space
+
+    - ``shifts`` -- shifts of the component rings of ``S``
+
+    This represents `\bigoplus_{j=1}^{t_i}H^0(\OO_{\PP^r}(-m^{(i)}_j))` for shifts `m^{(i)}_j`.
+
+    EXAMPLES::
+
+        sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+        sage: X = P2.subscheme([x^4 + y^4 + z^4])
+        sage: c = X.structure_sheaf()._cohomology
+        sage: c.cohomology_group_bottom(0)
+        Bottom Cohomology Group of dimension 1
+    """
+    def __init__(self, S, shifts):
+        """
+        Initialize.
+
+        TESTS::
+
+            sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+            sage: X = P2.subscheme([x^4 + y^4 + z^4])
+            sage: c = X.structure_sheaf(twist=1)._cohomology
+            sage: c.cohomology_group_bottom(0)
+            Bottom Cohomology Group of dimension 3
+        """
+        self.graded_ring = S
+        self.shifts = shifts
+
+        n = S.ngens()
+
+        basis = []
+        summands_basis = []
+        summands_index = []
+        rank = 0
+        for m in self.shifts:
+            # list of integer vectors whose entries are all non-negative integers and sum to -m
+            l = [vector(e) for e in IntegerListsLex(length=n, min_sum=-m, max_sum=-m)]
+            basis += l
+            summands_basis.append(l)
+            summands_index.append(rank)
+            rank += len(l)
+
+        self.summands_basis = summands_basis
+        self.summands_index = summands_index
+        self.basis = basis
+        self.vector_space = VectorSpace(S.base_ring(), rank)
+        self.rank = rank
+
+    def __repr__(self):
+        r"""
+        Return the string representation of ``self``.
+
+        EXAMPLES::
+
+            sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+            sage: X = P2.subscheme([x^4 + y^4 + z^4])
+            sage: c = X.structure_sheaf()._cohomology
+            sage: c.cohomology_group_bottom(0)
+            Bottom Cohomology Group of dimension 1
+        """
+        return f'Bottom Cohomology Group of dimension {self.rank}'
+
+
+class CohomologyGroupTop:
```

Comment:
This is just a copy of `CohomologyGroupBottom` except for the string representation and the integer vectors used. I believe. It would be better to combine them with an input of `is_top=boolean` that splits between the two cases. Or refactoring things to a common ABC.

Also, why aren't these subclasses of `Parent`, or at least `UniqueRepresentation`?

Reply:
They are two simple classes (in code length and the number of methods) each of which represent different mathematical objects. They are better to be separate. They are internal as the docstring of the module says. There is zero possibility that new method will be added to them in future. No need to make them technically complicated.

Please let them be left simple.

Reply:
The high level of duplication makes keeping them separate a fair amount of technical debt. I don't think they are simple enough to warrant this. They are not so distinct though. It's accessible through the API (as shown through the doctests), so it is not internal either. I think they should at least inherit from `UniqueRepresentation` and `SageObject`. I also think that they will never be extended is a bad assumption.

Reply:
I made them inherit from `SageObject`.

I want to keep the two classes separate as they represent different  objects, and there is nothing to make up a common superclass.

Reply:
That’s like saying we should have two classes for the even and odd integers. They are different mathematical objects, but as far as their uses, functionality, and implementation are concerned, they are essentially the same.

The code is very clear that there should be at least a common base class (you’re not disputing that there is a high level of duplication, right?), if not just having the output differ by an input (e.g., a boolean argument for top versus bottom).

Reply:
There are only two methods `__init__` and `__repr__`, and the two methods are different in the two classes. What high level of duplication is there? 

I guess that your hypothetical `EvenIntegers` and `OddIntegers` have over 10 (or say 3) methods and all of them are identical except `__init__` and `__repr__`. In that case, I would agree that there is a high level of duplication. 

Here `__init__` and `__repr__` are all, and they are not identical. I see no high level of duplication. 

Having the separate two classes, I could add an appropriate docstring to each class. Combining the two classes to one by using "a boolean argument for top versus bottom" is sacrificing readability for negligible code economy. I don't want to do that.

______

Fearing that you may be overestimating `CohomologyGroupTop` and `CohomologyGroupBottom` classes, I remark that the objects of the classes are not user-accessible at all. They are intermediate objects constructed and used during the computation and disappear.